### PR TITLE
Fix setting undefined so it returns an error and add tests.

### PR DESCRIPTION
**Pull request by mrister**
Newer redis version (post 3.0) changed handling of undefined value which caused problems and test failures.

This fix mitigates the problem.

### CODE REVIEW

**Comment by jkernech** [comment]
LGTM

**Comment by PuKoren** [comment]
I think this would break the compat with the https://github.com/BryanDonovan/node-cache-manager#specifying-what-to-cache and may lead to unexpected behavior

Default behavior seems to be ignoring undefined values, but can be overriden to store them anyway (if user make async code and want to store undefined as a valid value)

Also, in Redis node, it seems that undefined is considered as a valid value, but is transformed to a string: https://github.com/NodeRedis/node_redis#sending-commands

Since you can't `JSON.parse(undefined) `, this will indeed trigger an error (unlike `JSON.parse(null)`, valid).

`undefined` seems a correct value to be stored to me, if user wants to store it, he will have no way to do it with this change. Maybe we could check if it is 'undefined' before parsing as redis stores it this way? I know this is an overhead, but node version of redis chose to do it this way :/

What do you think about it?

**Comment by mrister** [comment]
@PuKoren  I had the same problem and thought about it a lot before issuing the PR. The one thing that led me here was the spec. The offending test was `'should not store an invalid value'`.

So per spec undefined was not saved, only change here is to actually get an error in .set callback (so the user does not need to assume he stored a value until seeing it later when calling get which in case of undefined returned a null.)

I see the point you make, and i like it even more then this solution, but in theory applying that would mean the need to change this test (spec), and this is a change in lib behaviour and ,again in theory,  per semver would require a new version release.

So not sure what to do here exactly... New version might seem the safest route to go without breaking anyone that even might be using this lib with an older redis version.

**Comment by PuKoren** [comment]
@mrister Not sure about the best way to go, overriding the spec looks good to me

**Comment by mrister** [comment]
@PuKoren  ok, let's go with the override then.

**Comment by mrister** [comment]
@PuKoren  Implemented the override thing, so undefined should be set if allowed by isCacheableValue function.

File: index.js [hunk 1]
```diff
@@ -71,7 +71,15 @@ function redisStore(args) {
       }
 
       if (opts.parse) {
-        result = JSON.parse(result);
+
+        try {
+          // allow undefined only if allowed by isCacheableValue
+          if(! ( (result === undefined || result === "undefined") && typeof args.isCacheableValue === 'function' && args.isCacheableValue(result))) {
```
Comment (PuKoren) [inline index.js]: I guess if result = `undefined` but `undefined` is not a cacheable value:

`! ((true) && false) === true` and you end up JSON.parsing undefined.


If value is not defined, you get `null` from redis: https://github.com/NodeRedis/node_redis/blob/master/test/commands/get.spec.js#L84

So null can be parsed, as well as "null", and you can't get undefined but only "undefined", and this can't be parsed.

What about:
```js
if (result === "undefined") {
  result = undefined;
} else {
  result = JSON.parse(result);
}
```

I think if result is not cacheable, it should not be stored and thus not be retrieved, so maybe we can skip this? What do you think about it?

Comment (mrister) [inline index.js]: Not sure what part to skip, the entire if section ?
Result can come as undefined and "undefined" and causes failure on JSON.parse
In case the value is not cacheable yes this will fail (if will allow to parse undefined), but this is expected and since it is in a try catch block you will get that error in your callback.
isCachebaleValue by default does not allow undefined.
There is a case here (referring to your node_redis link) where the key exists and value is 'undefined'.

Comment (PuKoren) [inline index.js]: On which case result can come with `undefined` value? I think it can come only with `"undefined"` (non-existing key in redis returns `null`). About the isCacheable, a non-cacheable value should not be stored in the first place, and we should retrieve then `null` no?

Comment (mrister) [inline index.js]: The function handleRespone is used for get and set, running our test suite showed both undefined and "undefined" need to be handled. I'll try to pinpoint it if needed.

Every store (as a plugin) can provide it's isCacheableValue but since it is a plugin to cache-manager, the cache manager should mandate if stuff that is not cacheable should be stored ( which it shouldn't as this is the whole point). If a value should not be stored, we should receive an error in callback (not swallow it) For example if we do a set(key, undefined) we (now with this PR) get an error since undefined is not allowed by default. I don't think it is good to allow a set to go without error and then get null (for the same key) as there is no indication of success nor failure and user can think the save succeeded, when it fact it did not.
The other edge case here is if the user allows undefined by isCachebaleValue and then this gets honored and value stored to db as "undefined" and retrieved as such.
I this PR we have test cases for both.

Comment (PuKoren) [inline index.js]: Ok I understand, this is used both in get and set. I was thinking it was only a get since there is some JSON.parsing.

Why do we parse it in a set?

Comment (mrister) [inline index.js]: Hm not sure, that was already there, I guess it was to have handleResponse as generic as possible and handle all cases.

Comment (PuKoren) [inline index.js]: hum ok, we will check for a cleanup later, because it can be used some thousands times/second we may don't want to check cacheableValues on each get, or parse objects on each set. thanks for the precisions

**Comment by PuKoren** [comment]
LGTM

**Comment by jkernech** [comment]
LGTM
